File: web/backend/Python-services/src/models/ModelVersion.py
"""
ModelVersion model for FAMS
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field
from ..constants import COLLECTIONS


class ModelVersion(BaseModel):
    """ModelVersion model representing versions of face recognition models"""
    modelId: int = Field(..., description="Unique identifier for the model version")
    modelName: str = Field(..., description="Name of the model")
    version: str = Field(..., description="Version number or identifier")
    deploymentDate: datetime = Field(default_factory=datetime.utcnow, description="Date when the model was deployed")
    description: Optional[str] = Field(None, description="Description of the model")
    deviceId: Optional[int] = Field(None, description="Reference to Device where model is deployed")
    checkpointPath: str = Field(..., description="Path to the model checkpoint file")
    status: str = Field(default="Active", description="Model status (Active or Inactive)")
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)
    isActive: bool = Field(default=True)
    collection: str = COLLECTIONS['MODEL_VERSION'] 

File: web/backend/NodeJS-services/database/models/ModelVersion.js
const mongoose = require('mongoose');
const { COLLECTIONS } = require('../constants');

/**
 * ModelVersion Schema
 * Represents a version of a face recognition model
 */
const ModelVersionSchema = new mongoose.Schema({
  modelId: {
    type: Number,
    required: true,
    unique: true,
    auto: true
  },
  modelName: {
    type: String,
    required: true
  },
  version: {
    type: String,
    required: true
  },
  deploymentDate: {
    type: Date,
    default: Date.now
  },
  description: {
    type: String
  },
  deviceId: {
    type: Number,
    ref: 'Device'
  },
  checkpointPath: {
    type: String,
    required: true
  },
  status: {
    type: String,
    enum: ['Active', 'Inactive'],
    default: 'Active'
  }
}, {
  timestamps: true,
  versionKey: false,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for getting device info
ModelVersionSchema.virtual('device', {
  ref: 'Device',
  localField: 'deviceId',
  foreignField: 'deviceId',
  justOne: true
});

// Virtual for getting face vectors using this model
ModelVersionSchema.virtual('faceVectors', {
  ref: 'FaceVector',
  localField: 'modelId',
  foreignField: 'modelId',
  justOne: false
});

module.exports = mongoose.model('ModelVersion', ModelVersionSchema, COLLECTIONS.MODEL_VERSION); 